galera/src/ist_proto.hpp

```diff
@@ -668,15 +668,27 @@ namespace galera
                             wsize = msg.len() - offset;
 
                             void*   const ptr(gcache_.malloc(wsize));
-                            ssize_t const r
-                                (socket.read(gu::AsioMutableBuffer(ptr, wsize)));
+                            // avoid allocating huge buffers, so receive it in chunks
```

Comment:
ptr is the pointer from GCache. So this is the pointer from mmap. We rely on signals to populate encryption cache pages. Unfortunately when ptr is passed to the socket (back to the syscall), signals are not notified. I didn't investigate why, but this mechanism simply doesn't work. So that's why we need this intermediate buffer (recv_buf) where we receive data from the socket and then memcpy to mmapped pointer.
Why in chunks? In general, there is no limit for the transfer size, the writeset can be huge, so we receive the data in 32kB chunks (we can increase chunk size if you think it is reasonable, but I think is fine).
Please note that this part of the code is used only during IST. "Normal" replication goes another way.